crates/trie/src/hashed_cursor/default.rs

```diff
@@ -1,54 +1,85 @@
-use super::{HashedAccountCursor, HashedCursorFactory, HashedStorageCursor};
+use super::{HashedCursor, HashedCursorFactory, HashedStorageCursor};
 use reth_db::{
     cursor::{DbCursorRO, DbDupCursorRO},
     tables,
     transaction::DbTx,
 };
-use reth_primitives::{Account, StorageEntry, B256};
+use reth_primitives::{Account, B256, U256};
 
 impl<'a, TX: DbTx> HashedCursorFactory for &'a TX {
     type AccountCursor = <TX as DbTx>::Cursor<tables::HashedAccounts>;
-    type StorageCursor = <TX as DbTx>::DupCursor<tables::HashedStorages>;
+    type StorageCursor =
+        DatabaseHashedStorageCursor<<TX as DbTx>::DupCursor<tables::HashedStorages>>;
 
     fn hashed_account_cursor(&self) -> Result<Self::AccountCursor, reth_db::DatabaseError> {
         self.cursor_read::<tables::HashedAccounts>()
     }
 
-    fn hashed_storage_cursor(&self) -> Result<Self::StorageCursor, reth_db::DatabaseError> {
-        self.cursor_dup_read::<tables::HashedStorages>()
+    fn hashed_storage_cursor(
+        &self,
+        hashed_address: B256,
+    ) -> Result<Self::StorageCursor, reth_db::DatabaseError> {
+        Ok(DatabaseHashedStorageCursor::new(
+            self.cursor_dup_read::<tables::HashedStorages>()?,
+            hashed_address,
+        ))
     }
 }
 
-impl<C> HashedAccountCursor for C
+impl<C> HashedCursor for C
 where
     C: DbCursorRO<tables::HashedAccounts>,
 {
-    fn seek(&mut self, key: B256) -> Result<Option<(B256, Account)>, reth_db::DatabaseError> {
+    type Value = Account;
+
+    fn seek(&mut self, key: B256) -> Result<Option<(B256, Self::Value)>, reth_db::DatabaseError> {
         self.seek(key)
     }
 
-    fn next(&mut self) -> Result<Option<(B256, Account)>, reth_db::DatabaseError> {
+    fn next(&mut self) -> Result<Option<(B256, Self::Value)>, reth_db::DatabaseError> {
         self.next()
     }
 }
 
-impl<C> HashedStorageCursor for C
+/// The structure wrapping a database cursor for hashed storage and
+/// a target hashed address. Implements [HashedCursor] and [HashedStorageCursor]
+/// for iterating over hashed storage.
+#[derive(Debug)]
+pub struct DatabaseHashedStorageCursor<C> {
+    cursor: C,
+    hashed_address: B256,
```

Comment:
could we get one line docs for the fields, I'd fine this easier to understand, even though the wrapper concept is easy to follow.